### ISSUE TITLE
'substr(...)' could be replaced with 'strpos(...)': False positive for string offset

### ISSUE THREAD

**Issue by donquixote** [kalessil/phpinspectionsea#1313]
| Subject        | Details                                                                       |
| :------------- | :---------------------------------------------------------------------------- |
| Plugin         | Php Inspections (EA Extended) v3.0.13 |
| Language level | PHP 5.6 |

## Code snippet
```php
/** @var string $delta */
if ($delta[0] === '#') {}
```

## Current inspection result
> 'strpos($delta, '#') === 0' can be used instead (improves maintainability).
from this inspection:
> 'substr(...)' could be replaced with 'strpos(...)'.

## Expected behaviour
No inspection warning, OR a separate inspection for this case, which can be separately enabled / disabled.

## Why?
I think it is a valid case that someone wants to keep their string offset lookup via `[$index]`, but replace their usages of `substr(...)`.

## Environment details
PhpStorm 2019.1.2 Preview
Build #PS-191.7141.5, built on April 24, 2019
JRE: 1.8.0_202-release-1483-b49 amd64
JVM: OpenJDK 64-Bit Server VM by JetBrains s.r.o
Linux 4.15.0-48-generic

**Comment by rentalhost** [comment]
In which case is better `$delta[0]` instead of `strpos()`? I ask it because `strpos()` is more secure than `$delta[0]`, that could causes an error if `$delta === ''`, for instance, while `strpos()` will works fine in all aspects.

**Comment by donquixote** [comment]
good point about `''`!
So when using `[0]` one should have an additional check for `''`. I think in most cases I actually did that in my code.

> In which case is better $delta[0] instead of strpos()?
If the string is really long and the character never occurs, it has to check every character in the string.
This also applies to strpos() vs substr() of course.

Either way I think it is two separate cases that deserve their own checkbox and explanation.

**Comment by rentalhost** [comment]
You are right, for very long cases `substr()` will be faster. I did a [benchmark](https://pastebin.com/Gbm39vi4) and I got this results:

```
Case indexWithEmptyStringSafe           duration of 0.258867s (avg. each 0.00000258s)
Case indexWithCharStringUnsafe          duration of 0.260480s (avg. each 0.00000260s)
Case indexWithEmptyStringSafeNegative   duration of 0.260535s (avg. each 0.00000260s)
Case indexWith16MStringUnsafe           duration of 0.268358s (avg. each 0.00000268s)
Case indexWithCharStringUnsafeNegative  duration of 0.268875s (avg. each 0.00000268s)
Case indexWith64KStringUnsafeNegative   duration of 0.271119s (avg. each 0.00000271s)
Case indexWith16MStringUnsafeNegative   duration of 0.271931s (avg. each 0.00000271s)
Case indexWith64KStringUnsafe           duration of 0.273311s (avg. each 0.00000273s)
Case substrWithCharStringNegative       duration of 0.358619s (avg. each 0.00000358s)
Case substrWith64KStringNegative        duration of 0.360468s (avg. each 0.00000360s)
Case substrWithEmptyString              duration of 0.363450s (avg. each 0.00000363s)
Case substrWith16MStringNegative        duration of 0.363657s (avg. each 0.00000363s)
Case strposWithEmptyString              duration of 0.363834s (avg. each 0.00000363s)
Case substrWithEmptyStringNegative      duration of 0.365460s (avg. each 0.00000365s)
Case substrWith64KString                duration of 0.366876s (avg. each 0.00000366s)
Case strposWithEmptyStringNegative      duration of 0.368112s (avg. each 0.00000368s)
Case substrWithCharString               duration of 0.373949s (avg. each 0.00000373s)
Case substrWith16MString                duration of 0.381116s (avg. each 0.00000381s)
Case indexWithEmptyStringSafeAt         duration of 0.382296s (avg. each 0.00000382s)
Case indexWithEmptyStringSafeAtNegative duration of 0.384522s (avg. each 0.00000384s)
Case strposWithCharStringNegative       duration of 0.385434s (avg. each 0.00000385s)
Case strposWithCharString               duration of 0.413507s (avg. each 0.00000413s)
Case strposWith64KString                duration of 0.816032s (avg. each 0.00000816s)
Case strposWith64KStringNegative        duration of 0.829365s (avg. each 0.00000829s)
Case strposWith16MStringNegative        duration of 183.627167s (avg. each 0.00183627s)
Case strposWith16MString                duration of 184.267251s (avg. each 0.00184267s)
```

So we have basically the same result for all cases (100.000 iterations each), except for very large strings using `strpos()` that is really very slow. But I belive that it should be treated as an exception. If you are working with strings with more than 64K in a looping, so maybe is better you silent the inspection. But note: even on worse case, it has consumed 0.001 second for each iteration. So I still should consider that as a micro-optimization.

**Comment by kalessil** [comment]
Checked, "normalization" to `strpos` is desired behavior in this case:
- covers non-empty strings
- enables a bunch of other checks/rules

**Comment by uuf6429** [comment]
Just found this GH issue - this behaviour has been bothering me for a while.

Why is `strpos` preferred over `substr($name, 0, 2) === $prefix` or `substr($name, -2) === $suffix`?
It's slower (as seen above) and less readable. The provided reason is also a bit ambiguous...what's "improves maintainability" supposed to mean (especially within the context of being less readable)?

@kalessil
